File: 509-fibonacci-number/509-fibonacci-number.cpp
class Solution {
public:
    int f(int idx, vector<int>& dp){
        
        if(idx <= 1) return idx;
        
        if(dp[idx] != -1) return dp[idx];
        
        return dp[idx] = f(idx-1, dp) + f(idx-2, dp);
    }
    
    int fib(int n) {
        
        // vector<int> dp(n+1, 0);
        
        if(n==0 || n==1) return n;
        
        // dp[0] = 0;
        int prev2 = 0;
        // dp[1] = 1;
        int prev1 = 1;
        
        for(int i=2; i<=n; i++){
            int curr = prev1 + prev2;
            prev2 = prev1;
            prev1 = curr;
        }
        
        return prev1;
    }
};

File: 509-fibonacci-number/509-fibonacci-number.py
class Solution:
    def f(self, idx, dp):
        if idx <= 1:
            return idx
        
        if dp[idx] != -1:
            return dp[idx]
        
        dp[idx] = self.f(idx-1, dp) + self.f(idx-2, dp)
        return dp[idx]
    
    
    def fib(self, n: int) -> int:
        if(n <= 1):
            return n
        
        first, second = 0, 1
        result=first+second
        for i in range(2, n+1):
            result = first + second
            first = second
            second = result
        return result